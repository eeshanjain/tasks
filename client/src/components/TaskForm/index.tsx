import React, { FC, useState } from 'react'

import { TaskFormPropsType } from '../../types'

const TaskForm: FC<TaskFormPropsType> = ({ addTask }) => {
  const [title, setTitle] = useState('')
  const [description, setDescription] = useState('')

  const addTaskHandler = () => addTask(title, description)
  const titleChangeHandler = (event: React.ChangeEvent<HTMLInputElement>) => setTitle(event.target.value)
  const descriptionChangeHandler = (event: React.ChangeEvent<HTMLInputElement>) => setDescription(event.target.value)

  return (
    <div>
      <input type="text" value={title} onChange={titleChangeHandler} placeholder="Title" />
      <input type="text" value={description} onChange={descriptionChangeHandler} placeholder="Description" />
      <button onClick={addTaskHandler}>Add Task</button>
    </div>
  )
}

export default TaskForm
