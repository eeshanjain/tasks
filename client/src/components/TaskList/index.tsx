import './taskList.css'

import React, { FC } from 'react'

import { TaskListPropsType } from '../../types'

const TaskList: FC<TaskListPropsType> = ({ tasks, editTask, deleteTask, markAsDone }) => {
  const sortedTasks = tasks.sort((task1, task2) => Number(task1.done) - Number(task2.done))

  const getClassStyle = (done: boolean) => done ? 'task-done' : ''

  return (
    <div>
      {sortedTasks.map(({_id, title, description, done}) => (
        <div key={_id}>
          <h3 className={getClassStyle(done)}>{title}</h3>
          <p className={getClassStyle(done)}>{description}</p>
          <button onClick={() => editTask(_id)}>Edit</button>
          <button onClick={() => deleteTask(_id)}>Delete</button>
          {!done && <button onClick={() => markAsDone(_id)}>Mark As Done</button>}
        </div>
      ))}
    </div>
  )
}

export default TaskList
