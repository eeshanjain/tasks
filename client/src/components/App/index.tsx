import * as Api from '../../api'

import React, { useEffect, useState } from 'react'

import { Task } from '../../types'
import TaskForm from '../TaskForm'
import TaskList from '../TaskList'

const App = () => {
  const [tasks, setTasks] = useState<Array<Task>>([])

  useEffect(() => {
    Api.getTasks()
      .then(res => {
        setTasks(res.data)
      })
      .catch(err => console.error(err))
  }, [])

  const addTask = (title: string, description: string) => {
    Api.addTask({ description, title })
      .then(() => window.location.reload())
      .catch(err => console.error(err))
  }

  const deleteTask = (taskId: string) => {
    Api.deleteTask(taskId)
      .then(() => window.location.reload())
      .catch(err => console.error(err))
  }

  const markAsDone = (taskId: string) => {
    Api.markTaskAsDone(taskId)
      .then(() => window.location.reload())
      .catch(err => console.error(err))
  }

  const editTask = () => {
    // eslint-disable-next-line no-alert
    alert('Edit task')
  }

  return (
    <div>
      <h1>Task Manager</h1>
      <TaskForm addTask={addTask} />
      <TaskList tasks={tasks} deleteTask={deleteTask} editTask={editTask} markAsDone={markAsDone} />
    </div>
  )
}

export default App
