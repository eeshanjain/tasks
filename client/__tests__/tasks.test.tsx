import App from '../src/components/App'
import React from 'react'
import TaskForm from '../src/components/TaskForm'
import TaskList from '../src/components/TaskList'
import renderer from 'react-test-renderer'

test('should render App', () => {
  const component = renderer.create(<App />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('should render TaskForm', () => {
  const component = renderer.create(<TaskForm addTask={jest.fn} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('should render TaskList', () => {
  const tasks = [
    {
      _id: '1',
      description: 'test description 1',
      done: false,
      title: 'test title 1',
    },
    {
      _id: '2',
      description: 'test description 2',
      done: true,
      title: 'test title 2'
    }
  ]
  const component = renderer.create(<TaskList tasks={tasks} editTask={jest.fn} deleteTask={jest.fn} markAsDone={jest.fn} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
