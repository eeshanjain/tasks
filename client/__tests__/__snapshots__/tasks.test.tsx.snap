// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render App 1`] = `
<div>
  <h1>
    Task Manager
  </h1>
  <div>
    <input
      onChange={[Function]}
      placeholder="Title"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
    >
      Add Task
    </button>
  </div>
  <div />
</div>
`;

exports[`should render TaskForm 1`] = `
<div>
  <input
    onChange={[Function]}
    placeholder="Title"
    type="text"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value=""
  />
  <button
    onClick={[Function]}
  >
    Add Task
  </button>
</div>
`;

exports[`should render TaskList 1`] = `
<div>
  <div>
    <h3
      className=""
    >
      test title 1
    </h3>
    <p
      className=""
    >
      test description 1
    </p>
    <button
      onClick={[Function]}
    >
      Edit
    </button>
    <button
      onClick={[Function]}
    >
      Delete
    </button>
    <button
      onClick={[Function]}
    >
      Mark As Done
    </button>
  </div>
  <div>
    <h3
      className="task-done"
    >
      test title 2
    </h3>
    <p
      className="task-done"
    >
      test description 2
    </p>
    <button
      onClick={[Function]}
    >
      Edit
    </button>
    <button
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
</div>
`;
